### DISCUSSION TITLE
yes I also have linking issues

### DISCUSSION THREAD

**Discussion by UltraBlackLinux** [ocornut/imgui#7951] [New Users Build/Link/Run issues ONLY!]
```
clang++ main.cpp -I ./imgui -l SDL2 -o /tmp/osm.x
/usr/bin/ld: /tmp/main-375907.o: in function `main':
main.cpp:(.text+0x10d): undefined reference to `ImGui::DebugCheckVersionAndDataLayout(char const*, unsigned long, unsigned long, unsigned long, unsigned long, unsigned long, unsigned long)'
/usr/bin/ld: main.cpp:(.text+0x116): undefined reference to `ImGui::CreateContext(ImFontAtlas*)'
/usr/bin/ld: main.cpp:(.text+0x11b): undefined reference to `ImGui::GetIO()'
/usr/bin/ld: main.cpp:(.text+0x143): undefined reference to `ImGui_ImplSDL2_InitForOpenGL(SDL_Window*, void*)'
/usr/bin/ld: main.cpp:(.text+0x14c): undefined reference to `ImGui_ImplOpenGL3_Init(char const*)'
/usr/bin/ld: main.cpp:(.text+0x15e): undefined reference to `ImGui_ImplSDL2_ProcessEvent(SDL_Event const*)'
/usr/bin/ld: main.cpp:(.text+0x163): undefined reference to `ImGui_ImplOpenGL3_NewFrame()'
/usr/bin/ld: main.cpp:(.text+0x168): undefined reference to `ImGui_ImplSDL2_NewFrame()'
/usr/bin/ld: main.cpp:(.text+0x16d): undefined reference to `ImGui::NewFrame()'
/usr/bin/ld: main.cpp:(.text+0x176): undefined reference to `ImGui::ShowDemoWindow(bool*)'
/usr/bin/ld: main.cpp:(.text+0x17b): undefined reference to `ImGui::Render()'
/usr/bin/ld: main.cpp:(.text+0x180): undefined reference to `ImGui::GetDrawData()'
/usr/bin/ld: main.cpp:(.text+0x188): undefined reference to `ImGui_ImplOpenGL3_RenderDrawData(ImDrawData*)'
clang++: error: linker command failed with exit code 1 (use -v to see invocation)
```
I don't see what's wrong...
Autocompletion works alright and that's generated via `compile_commands.json` so it's definitely able to find the library, but the darn linker is not doing its thing.

Any ideas?
Thanks!

**Comment by ocornut** [comment]
The linker is doing its thing. You’re just not linking with anything else than the compiled output of main.cpp. Someone will probably come up with a longer answer but you need to compile imgui sources and link them, and its backends as well.

**Reply by UltraBlackLinux** [reply]
right, I was wondering about that. For some reason I thought that clang would automatically compile included dependencies as well...

Managed to get it to build like this: `bear -- clang++ imgui/*.cpp imgui/backends/imgui_impl_opengl3.cpp imgui/backends/imgui_impl_sdl2.cpp main.cpp -I ./imgui -I /usr/include/SDL2/ -l SDL2 -o /tmp/osm.x`

Thanks!

**Reply by ocornut** [reply]
This is one way and it’ll work but means you are recompiling the entirety of dear imgui everytime you build your project. Read about compiling sources to objets files and linking objects files. Most people use build systems such as Makefike, cmake, premake, visual studio’s msbuild to solve this problem.

**Reply by UltraBlackLinux** [reply]
that is a valid argument but I'm not really trying to go super fast. I just want it to work ;)
